pkg/helpers/crd.go

```diff
@@ -102,3 +123,78 @@ func (s SecretSortList) Less(i, j int) bool {
 
 	return s.Items[i].CreationTimestamp.Before(&s.Items[j].CreationTimestamp)
 }
+
+// Checks if the slice of strings contains a specific string
+func SliceContainsString(list []string, s string) bool {
+	for _, v := range list {
+		if v == s {
+			return true
+		}
+	}
+	return false
+}
+
+// handleDeletion will handle the finalizers if the object was deleted.
+// It will return true, if the deletiontimestamp is set. If the object was
+// removed the reconcile can be returned.
+// The resource to be removed can be nil, in case there's no direct resource,
+// like for a cluster without Vault.
+func HandleDeletion(instance metav1.Object, finalizerName string, client client.Client) (bool, error) {
+	if instance.GetDeletionTimestamp().IsZero() {
+		return false, nil
+	}
+
+	annotationValue, exists := instance.GetAnnotations()[DeleteProtectionAnnotation]
+
+	var protected bool
+	var err error
+	if exists {
+		protected, err = strconv.ParseBool(annotationValue)
+		// Assume true if it can't be parsed
+		if err != nil {
+			protected = true
+			// We need to reset the error again, so we don't trigger any unwanted side effects...
+			err = nil
+		}
+	} else {
+		protected = false
+	}
+
+	if SliceContainsString(instance.GetFinalizers(), finalizerName) && !protected {
+
+		controllerutil.RemoveFinalizer(instance, finalizerName)
+		obj, ok := instance.(runtime.Object)
+		if !ok {
+			return false, fmt.Errorf("can't convert object to runtime.instance")
+		}
+
+		// if this fails the finalizer will remain, effectively not deleting the CR
+		err = client.Update(context.TODO(), obj)
```

Comment:
I don't see the benefit of that. If the finalizers are removed, the CR is effectively deleted, so the reconcile function should only cleanup and return.

Also if the reconcile function should do that there's code duplication and stuff that might get lost when implementing a new controller.

Reply:
hmm true. The object should be gone if the update was successful.

Reply:
So now if the finalizer is removed it will return true so the reconcile knows that it should update the object and remove the external resources. Thus deleting it (if all finalizers are gone).